### DISCUSSION TITLE
[Announcement] more explanation on safety property

### DISCUSSION THREAD

**Discussion by KihongHeo** [prosyslab-classroom/cs424-program-reasoning#350] [Announcements]
Hi,

Let me clarify one thing regarding the question from Sanga (상아) during the lecture today.
The proof of the safety property of an algorithm can be done as follows.

For simplicity, let us consider a search algorithm rather than a sorting algorithm. Consider a linear search algorithm: given an array `a`, lower bound index `l`, upper bound index `u`, and target `e`, the following function checks whether there exists `e` in the array within the boundaries:

```c
bool LinearSearch(int []a, int l, int u, int e) {
  int i = l;
  while (i <= u) {
    if (a[i] == e) return true;
    i = i + 1;
  }
  return false;
}
```

To show the safety property "the algorithm returns true if there exists `e`", it is enough to show the following is a loop invariant:
```
l <= i /\ (forall  j. l <= j < i  -> a[j] != e)
```
This is the "good" state we want our program remains in during the execution of the loop.
Why? If it is proven to be a loop invariant, it is fairly easy to prove the correctness afterward. I recommend you prove it by yourself as an exercise. 

We will revisit this example in a few weeks.

**Comment by oojahooo** [comment]
I think the loop invariant expression has a typo. Maybe the correct one is `l <= i /\ (forall j. l <= j < i -> a[j] != e)`.

**Comment by KihongHeo** [comment]
Thanks @oojahooo . Let me fix.
